src/amy.c

```diff
@@ -425,32 +426,63 @@ void amy_add_event_internal(struct event e, uint16_t base_osc) {
     if(AMY_IS_SET(e.wave)) { d.param=WAVE; d.data = *(uint32_t *)&e.wave; add_delta_to_queue(d); }
     if(AMY_IS_SET(e.patch)) { d.param=PATCH; d.data = *(uint32_t *)&e.patch; add_delta_to_queue(d); }
     if(AMY_IS_SET(e.midi_note)) { d.param=MIDI_NOTE; d.data = *(uint32_t *)&e.midi_note; add_delta_to_queue(d); }
-    for (int i = 0; i < NUM_COMBO_COEFS; ++i)
-        if(AMY_IS_SET(e.amp_coefs[i])) {  d.param=AMP + i; d.data = *(uint32_t *)&e.amp_coefs[i]; add_delta_to_queue(d); }
+    
 
-    for (int i = 0; i < NUM_COMBO_COEFS; ++i) {
-        if(AMY_IS_SET(e.freq_coefs[i])) {
+    // To match the string parser's behavior, we should set all coefs to 0 if they are not set, if any one of them is set.
+    // See https://github.com/shorepine/amy/issues/163
+
+    setflag = 0;
+    for (int i = 0; i < NUM_COMBO_COEFS; ++i) if(AMY_IS_SET(e.amp_coefs[i])) setflag = 1;
+    if(setflag) {
+        for (int i = 0; i < NUM_COMBO_COEFS; ++i) {
+            if(AMY_IS_UNSET(e.amp_coefs[i])) e.amp_coefs[i] = 0;
+            d.param=AMP + i; 
+            d.data = *(uint32_t *)&e.amp_coefs[i]; 
+            add_delta_to_queue(d);
+        }
+    }
+    
+    setflag = 0;
+    for (int i = 0; i < NUM_COMBO_COEFS; ++i) if(AMY_IS_SET(e.freq_coefs[i])) setflag = 1;
+    if(setflag) {
+        for (int i = 0; i < NUM_COMBO_COEFS; ++i) {
+            if(AMY_IS_UNSET(e.freq_coefs[i])) e.freq_coefs[i] = 0;
             float freq_coef = e.freq_coefs[i];
             // Const freq coef is in Hz, rest are linear.
             if (i == COEF_CONST) freq_coef = logfreq_of_freq(freq_coef);
             d.param=FREQ + i; d.data = *(uint32_t *)&freq_coef; add_delta_to_queue(d);
         }
     }
 
-    if(AMY_IS_SET(e.filter_freq_coefs[COEF_CONST])) { float filter_logfreq = logfreq_of_freq(e.filter_freq_coefs[COEF_CONST]); d.param=FILTER_FREQ; d.data = *(uint32_t *)&filter_logfreq; add_delta_to_queue(d); }
```

Comment:
This was weird and bears a closer looking at -- we were checking filter_freq_coefs[COEF_CONST] twice, but doing the same thing each time. I have it just looking at it once now

Reply:
Yes, you're right.  `[filter_]freq_coefs[COEF_CONST]` is a special case as it is converted from Hz to logfreq units, unlike the remaining parameters.  But I'd put the special case in twice.  I like your resolution.

However, I disagree with clearing unset values to 0 (the remainder of this PR).  I think the C interface should support setting individual components of the `_coefs` vectors one at a time.  We pack them into a vector for the Python interface for brevity and to save on wire-protocol symbols, but there are plenty of use-cases for changing them individually (this is exactly what you want to happen when you move the VCF Freq slider on your MIDI controller).